### PR TITLE
Define the Quiz Data Model

### LINKED ISSUE #33: Define the Quiz Data Model
As an efficient developer team, we want to know how the quiz data will look, in order to work simultaneously on our tasks.
Acceptance Criteria

- [x] A Question class is created (take milestone 5 into consideration)
  - [x] Contains the question, explanation, answers, correct answer index, and categories.
- [x] A UserAnswer class is created
  - [x] Contains the given answer index, a correct/wrong flag
- [x] A Category class is created
  - [x] Contains the category name and a flag if it is a special category
  - [x] Review, exam, and all are special categories
- [x] A QuizState class is created (define method stubs for changing state as well)
  - [x] Contains a flag whether the user is currently taking a quiz
  - [x] Contains the questions (simply hardcoded for now), user answers, score
  - [x] Contains helper methods for the ui (nr of questions, current question nr, etc.)
- [x] All developers agree on the data model and a wiki page is created
  - [x] The MqttCar class is added (static and @Override attributes can be ignored)
  - [x] The CarState class is added (private and static attributes can be ignored)
  - [x] The Question class is added (private and static attributes can be ignored)
  - [x] The UserAnswer class is added (private and static attributes can be ignored)
  - [x] The QuizState class is added (private and static attributes can be ignored)

### PR DESCRIPTION
**Pull request by plepisnew**
Closes #33 

Created Models for Question, Category, UserAnswer and QuizState.
Encapsulated fields for each Model. Added Helper methods that will be necessary in the UI.

Note: Inconsistency between schema in [Wiki page](https://github.com/DIT113-V22/group-01/wiki/8.-Database-for-Quiz-Questions#quiz-question-table) and [Acceptance Criteria](https://github.com/DIT113-V22/group-01/issues/33)

### CODE REVIEW

**Comment by platisd** [comment]
Why is `33` in the title?
That is an unconventional way of referring to issues.

**Comment by plepisnew** [comment]
> Why is `33` in the title? That is an unconventional way of referring to issues.

Should I also refer to the issue in the title, using `#33` ? I just assumed we only have to refer to it in the description.

**Comment by Lancear** [comment]
> > Why is `33` in the title? That is an unconventional way of referring to issues.
> 
> Should I also refer to the issue in the title, using `#33` ? I just assumed we only have to refer to it in the description.

I think you can simply remove the 33 from the title since its in the description already
image

**Comment by plepisnew** [comment]
> > > Why is `33` in the title? That is an unconventional way of referring to issues.
> > 
> > 
> > Should I also refer to the issue in the title, using `#33` ? I just assumed we only have to refer to it in the description.
> 
> I think you can simply remove the 33 from the title since its in the description already image

will do! thanks

**Comment by platisd** [comment]
What @Lancear said 👍

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/model/Question.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
+package com.example.smartcarmqttapp.model;
+import java.util.List;
+
+// Class for creating each possible question and the answers & explanations to them
+public class Question {
+
+    private int id;
```
Comment (JohnWebb00) [inline SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/model/Question.java]: Remove id variable. Id is assigned on table creation

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/model/Question.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    private String secondAnswer;
+    private String thirdAnswer;
+    private String fourthAnswer;
+    private int correctIndex;
+    private Category category;
+    private boolean needsReview;
+
+    // Constructor with all fields
+    public Question(
+            int id,
```
Comment (JohnWebb00) [inline SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/model/Question.java]: Remove id variable. Id is assigned on table creation

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/model/Question.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            String explanation,
+            String firstAnswer,
+            String secondAnswer,
+            String thirdAnswer,
+            String fourthAnswer,
+            int correctIndex,
+            Category category,
+            boolean needsReview
+    ) {
+        this.id = id;
```
Comment (JohnWebb00) [inline SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/model/Question.java]: same here

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/model/Question.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        this.secondAnswer = secondAnswer;
+        this.thirdAnswer = thirdAnswer;
+        this.fourthAnswer = fourthAnswer;
+        this.correctIndex = correctIndex;
+        this.category = category;
+        this.needsReview = needsReview;
+    }
+
+    // Constructor without id
+    public Question(
```
Comment (JohnWebb00) [inline SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/model/Question.java]: Just need this constructor and the empty constructor

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/model/Question.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    private int id;
+    private String question;
+    private String explanation;
+    private String firstAnswer;
+    private String secondAnswer;
+    private String thirdAnswer;
+    private String fourthAnswer;
+    private int correctIndex;
+    private Category category;
+    private boolean needsReview;
```
Comment (JohnWebb00) [inline SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/model/Question.java:15]: needsReview is a good edition. will add to my branch as well

**Review by JohnWebb00 — CHANGES_REQUESTED** [review]

**Review by JohnWebb00 — APPROVED** [review]

**Pull request merged**